trieste/space.py

```diff
@@ -480,6 +503,186 @@ def __eq__(self, other: object) -> bool:
+            categories = [categories]  # type: ignore[assignment]
+
+        if not isinstance(categories, Sequence) or not (
+            all(
+                isinstance(x, Sequence)
+                and not isinstance(x, str)
+                and all(isinstance(y, str) for y in x)
+                for x in categories
+            )
+            or all(isinstance(x, int) for x in categories)
+        ):
+            raise TypeError(
+                f"Invalid category description {categories!r}: " "expected either numbers or names."
+            )
+
+        elif any(isinstance(x, int) for x in categories):
+            category_lens: Sequence[int] = categories  # type: ignore[assignment]
+            if any(x <= 0 for x in category_lens):
+                raise ValueError("Numbers of categories must be positive")
+            tags = [tuple(f"{i}" for i in range(n)) for n in category_lens]
+        else:
+            category_names: Sequence[Sequence[str]] = categories  # type: ignore[assignment]
+            if any(len(ts) == 0 for ts in category_names):
+                raise ValueError("Category name lists cannot be empty")
+            tags = [tuple(ts) for ts in category_names]
+
+        self._tags = tags
+
+        ranges = [tf.range(len(ts), dtype=dtype) for ts in tags]
+        meshgrid = tf.meshgrid(*ranges, indexing="ij")
+        points = (
+            tf.reshape(tf.stack(meshgrid, axis=-1), [-1, len(tags)]) if tags else tf.zeros([0, 0])
+        )
+
+        super().__init__(points)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"CategoricalSearchSpace({self._tags!r})"
+
+    @property
+    def has_bounds(self) -> bool:
+        return False
+
+    @property
+    def lower(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def upper(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def tags(self) -> Sequence[Sequence[str]]:
+        """The tags of the categories."""
+        return self._tags
+
+    @property
+    def one_hot_encoder(self) -> EncoderFunction:
+        """A one-hot encoder for the numerical indices."""
+
+        def encoder(x: TensorType) -> TensorType:
+            flat_x, unflatten = flatten_leading_dims(x)
+            if flat_x.shape[-1] != len(self.tags):
+                raise ValueError(
+                    "Invalid input for one-hot encoding: "
+                    f"expected {len(self.tags)} tags, got {flat_x.shape[-1]}"
+                )
+            columns = tf.split(flat_x, flat_x.shape[-1], axis=1)
+            encoders = [
+                tf.keras.layers.CategoryEncoding(num_tokens=len(ts), output_mode="one_hot")
+                for ts in self.tags
+            ]
+            encoded = tf.concat(
+                [encoder(column) for encoder, column in zip(encoders, columns)], axis=1
+            )
```

Comment:
Given `x` is now a float type, do we need to do some checking/casting similar to `to_tags`? I'm not sure if `tf.keras.layers.CategoryEncoding` allows non-integer inputs.

Reply:
It's only the underling self.points in GeneralDiscreteSearchSpace that are now floats. The one hot encoding uses self.tags directly so isn't affected. (And this is all tested.)